File: sorters_c/sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void print_array(int arr[], int length)
{
	int i;
	for (i = 0; i < length; i++)
	{
		printf("%d\n", arr[i]);
	}
}

void insertion_sort(int arr[], int length)
{
	int i, j, key;
	for (i = 0; i < length; i++)
	{
		j = i - 1;
		key = arr[i];

		while (j >= 0 && arr[j] > key)
		{
			arr[j + 1] = arr[j];
			j = j - 1;
		}
		arr[j + 1] = key;
	}
}

void merge(int input[], int helper[], int length, int first, int mid, int last)
{
	for (int i = first; i <= last; i++)
	{
		helper[i] = input[i];
	}

	int inFirst = first;
	int inSecond = mid + 1;
	for (int inNew = first; inNew <= last; inNew++)
	{
		if (inFirst > mid)
		{
			input[inNew] = helper[inSecond];
			inSecond++;
		}
		else if (inSecond > last)
		{
			input[inNew] = helper[inFirst];
			inFirst++;
		}
		else
		{
			if (helper[inFirst] < helper[inSecond])
			{
				input[inNew] = helper[inFirst];
				inFirst++;
			}
			else
			{
				input[inNew] = helper[inSecond];
				inSecond++;
			}
		}
	}
}

void merge_sort(int input[], int helper[], int length, int first, int last)
{
	if (first < last)
	{
		int mid = first + (last - first) / 2;
		merge_sort(input, helper, length, first, mid);
		merge_sort(input, helper, length, mid + 1, last);
		merge(input, helper, length, first, mid, last);
	}
}

void load_array(FILE *in_file, int arr[])
{
	rewind(in_file);

	int i = 0;
	int ctr = 0;
	while (!feof(in_file))
	{
		fscanf(in_file, "%d", &i);
		arr[ctr] = i;
		ctr++;
	}
}

int main(void)
{
	int size = 100000;

	FILE *in_file;
	char *in_filename;
	in_filename = "../data.txt";
	in_file = fopen(in_filename, "r");

	FILE *out_file;
	char *out_filename;
	out_filename = "../output_c.txt";
	out_file = fopen(out_filename, "w");

	struct timespec start, end;
	int dif;

	int arr[size];

	fprintf(out_file, "nbr, time in ns\n");

	for (int k = 1; k <= 1; k++)
	{
		load_array(in_file, arr);

		clock_gettime(CLOCK_MONOTONIC, &start);
		
		insertion_sort(arr, size);
        
        	// int helper[size];
        	// merge_sort(arr, helper, size, 0, size - 1);

		clock_gettime(CLOCK_MONOTONIC, &end);
		dif = (end.tv_nsec - start.tv_nsec);
		fprintf(out_file, "%d, %d\n", k, dif);
	}

	fclose(in_file);
	fclose(out_file);

	return 0;
}

File: sorters_java/src/uaps/Sort.java
package uaps;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class Sort {
    public static void insertionSort(int[] arr) {
        for (int i = 0; i < arr.length; i++) {
            int key = arr[i];
            int j = i - 1;
            while (j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j = j - 1;
            }
            arr[j + 1] = key;
        }
    }

    public static void merge(int[] input, int[] helper, int first, int mid, int last) {
        for (int i = first; i <= last; i++) {
            helper[i] = input[i];
        }

        int inFirst = first;
        int inSecond = mid + 1;
        for (int inNew = first; inNew <= last; inNew++) {
            if (inFirst > mid) {
                input[inNew] = helper[inSecond];
                inSecond++;
            }

            else if (inSecond > last) {
                input[inNew] = helper[inFirst];
                inFirst++;
            }

            else {
                if (helper[inFirst] < helper[inSecond]) {
                    input[inNew] = helper[inFirst];
                    inFirst++;
                }

                else {
                    input[inNew] = helper[inSecond];
                    inSecond++;
                }
            }
        }
    }

    public static void mergeSort(int[] input, int[] helper, int first, int last) {
        if (first < last) {
            int mid = first + (last - first) / 2;
            mergeSort(input, helper, first, mid);
            mergeSort(input, helper, mid + 1, last);
            merge(input, helper, first, mid, last);
        }
    }

    public static int[] loadFromFile(String fileName) throws FileNotFoundException {
        File input = new File(fileName);
        Scanner inScanner = new Scanner(input);

        int[] arr = new int[100000];
        int i = 0;
        while (inScanner.hasNext()) {
            arr[i] = (Integer.parseInt(inScanner.nextLine()));
            i++;
        }
        inScanner.close();
        return arr;
    }

    public static void main(String[] args) throws IOException {
        FileWriter outWriter = new FileWriter(args[1], false);
        outWriter.write("nbr, time in s\n");

        for (int i = 1; i <= Integer.parseInt(args[2]); i++) {
            int[] a = loadFromFile(args[0]);
            long t0 = System.nanoTime();

            // insertionSort(a);
            int[] b = new int[a.length];
            mergeSort(a, b, 0, a.length - 1);

            long t1 = System.nanoTime();
            double timeTaken = (t1 - t0) / 1e9;
            outWriter.write(i + ", " + timeTaken + "\n");
        }
        outWriter.close();
    }
}
